internal/pkg/agent/application/dispatcher/dispatcher.go

```diff
@@ -290,86 +396,64 @@ func (ad *ActionDispatcher) scheduleRetry(ctx context.Context, action fleetapi.R
 	}
 }
 
-func (ad *ActionDispatcher) handleExpired(
-	expired []fleetapi.Action,
-	upgradeDetailsSetter details.Observer) {
-
-	for _, e := range expired {
-		if e.Type() == fleetapi.ActionTypeUpgrade {
-			// there is a scheduled upgrade set, if it isn't the same actions as
-			// the expired, the current status take precedence
-			if ad.lastUpgradeDetails != nil &&
-				ad.lastUpgradeDetails.ActionID != e.ID() {
-				continue
-			}
-
-			version := "unknown"
-			expiration := "unknown"
-			if upgrade, ok := e.(*fleetapi.ActionUpgrade); ok {
-				version = upgrade.Data.Version
-				expiration = upgrade.ActionExpiration
-			}
-			ad.lastUpgradeDetails = details.NewDetails(version, details.StateFailed, e.ID())
-			ad.lastUpgradeDetails.Fail(fmt.Errorf("upgrade action %q expired on %s",
-				e.ID(), expiration))
-
-			upgradeDetailsSetter(ad.lastUpgradeDetails)
-		}
-	}
-}
+// GetScheduledUpgradeDetails returns the upgrade details of the next upgrade action, if any. It also adjusts accordingly
+// the upgrade details if the action has expired.
+func GetScheduledUpgradeDetails(log *logger.Logger, actions []fleetapi.ScheduledAction, ts time.Time) *details.Details {
+	var nextUpgradeAction *fleetapi.ActionUpgrade
+	var nextUpgradeStartTime time.Time
+	var nextUpgradeExpirationTime time.Time
 
-func (ad *ActionDispatcher) reportNextScheduledUpgrade(input []fleetapi.Action, detailsSetter details.Observer, log *logger.Logger) {
-	var nextUpgrade *fleetapi.ActionUpgrade
-	for _, action := range input {
-		sAction, ok := action.(fleetapi.ScheduledAction)
-		if !ok {
-			continue
-		}
-
-		uAction, ok := sAction.(*fleetapi.ActionUpgrade)
+	for _, queuedAction := range actions {
+		upgradeAction, ok := queuedAction.(*fleetapi.ActionUpgrade)
 		if !ok {
 			continue
 		}
 
-		start, err := uAction.StartTime()
+		// all queued upgrade actions must have a start time
+		upgradeStartTime, err := upgradeAction.StartTime()
 		if err != nil {
-			log.Errorf("failed to get start time for scheduled upgrade action [id = %s]", uAction.ID())
+			log.Errorf("failed to get start time for scheduled upgrade action [id = %s]: %v", upgradeAction.ID(), err)
 			continue
 		}
 
-		if !start.After(time.Now()) {
-			continue
-		}
-
-		if nextUpgrade == nil {
-			nextUpgrade = uAction
-			continue
+		// not all scheduled upgrade actions have an expiration, e.g. a scheduled retried upgrade because it failed
+		// does not have an expiration
+		upgradeExpirationTime, err := upgradeAction.Expiration()
+		if err != nil {
+			if !errors.Is(err, fleetapi.ErrNoExpiration) {
+				// this is not a non-expiring upgrade action
+				log.Errorf("failed to get expiration time for scheduled upgrade action [id = %s]: %v", upgradeAction.ID(), err)
```

Comment:
Considering not all scheduled upgrade actions are expected to have an expiration, should we make this a warning instead of an error?

Reply:
Never mind, I got confused by the error message.  I see you are logging an error that occurred when getting the expiration time, but only when the action is _expected_ not to have an expiration.  The error is not that the action has no expiration time.